compiler/rustc_mir/src/const_eval/machine.rs

```diff
@@ -37,6 +38,10 @@ impl<'mir, 'tcx> InterpCx<'mir, 'tcx, CompileTimeInterpreter<'mir, 'tcx>> {
         if instance.def.requires_caller_location(self.tcx()) {
             return Ok(false);
         }
+        // only memoize instrinsics
+        if !matches!(instance.def, InstanceDef::Intrinsic(_)) {
```

Comment:
I'm okay with leaving the cleanup to a future PR.

> The new intrinsic has non-ZST args, so it doesn't have this problem until we also memoize function calls with arguments.

Ah, good point. @vn-ki please add a comment explaining this.

Reply:
There is a comment below which says

>         // For the moment we only do this for functions which take no arguments
>        // (or all arguments are ZSTs) so that we don't memoize too much.

Is that enough or should I add one more?

Reply:
I meant a comment specifically about the allocation intrinsic, because it is the only impure intrinsic (but it happens to also have a non-ZST argument).

Reply:
Add an additional comment specific to the fact that it would be *unsound* to memoize any function which returns pointers to the heap (and possibly even just interacts with the heap)